### ISSUE TITLE
Flutter autocomplete suggestions not working

### ISSUE THREAD

**Issue by BojanDolic** [flutter/flutter-intellij#5978]
My autocomplete suggestions just stopped working.
As an example, none of the widgets show inside suggestions window when I start typing, here is the screenshot when I want to add BoxDecoration:

[First image](https://i.imgur.com/hDbqncK.png)
[Second image](https://i.imgur.com/Myg8DMt.png)

## Steps to Reproduce

I'm not exactly sure how this happened, but the issue started when I ran android emulator inside tool window ( I ran it as a separate window before.)

I created this project few days ago and everything was working fine until now.

## Version info

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 2.10.1, on Microsoft Windows [Version 10.0.19043.1526], locale sr-Latn-BA)
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    X cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.
[√] Chrome - develop for the web
[X] Visual Studio - develop for Windows
    X Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components
[√] Android Studio (version 2020.3)
[√] Android Studio (version 2021.1)
[√] IntelliJ IDEA Community Edition (version 2021.1)
[√] Connected device (4 available)
[√] HTTP Host Availability

! Doctor found issues in 2 categories.
```

**Comment by stevemessick** [comment]
`BoxDecoration` is part of Material Design. Do you have `uses-material-design: true` in your `pubspec.yaml`?

**Comment by BojanDolic** [comment]
> `BoxDecoration` is part of Material Design. Do you have `uses-material-design: true` in your `pubspec.yaml`?

Yes I do. Everything was working correctly until now. Autocomplete for widgets worked fine until now.
I tried restarting my PC also which didn't help.

I also tried autocomplete in another project and it worked fine. I created empty project and autocomplete also worked so It seems that the issue is tied to this project but what is bothering me is that everything was ok until I launched emulator. It seems I can't find a way to fix this.

**Comment by stevemessick** [comment]
I would not expect the emulator to affect completion. To be honest, this is a puzzle. I don't know what project-specific thing could cause this. Your quickest remedy may be to re-create your project. If you don't want to do that you could try `Flutter Pub Get` and `Flutter Pub Upgrade` from the Tools > Flutter menu.

Have you ever opened the Android module as its own project in Android Studio (`Tools > Flutter > Open Android module in Android Studio`)? That could possibly cause changes (rarely).

**Comment by BojanDolic** [comment]
> I would not expect the emulator to affect completion. To be honest, this is a puzzle. I don't know what project-specific thing could cause this. Your quickest remedy may be to re-create your project. If you don't want to do that you could try `Flutter Pub Get` and `Flutter Pub Upgrade` from the Tools > Flutter menu.

Luckily project is not big so I recreated it. I tried `Flutter Pub Get` and `Flutter Pub Upgrade` with no success.

> Have you ever opened the Android module as its own project in Android Studio (`Tools > Flutter > Open Android module in Android Studio`)? That could possibly cause changes (rarely).

I did not. When I go to `Tools > Flutter` there is no `Open Android module in Android` option shown.

<img src="https://i.imgur.com/MHTt5Ii.png" />

**Comment by stevemessick** [comment]
Sorry, but the menu item name changes depending on what is selected in the Project view. In your screen shot, it is named `Open for Editing in Android Studio`. Have you used that?

Does completion work as expected in the new project?

Since I'm guessing the answers to these two questions are "no" and "yes", I'm going to see if I can get some help trying to understand what went wrong in the original project.
@alexander-doroshko @bwilkerson can you think of anything that might explain the original problem?

**Comment by bwilkerson** [comment]
I can think of a couple of possibilities, but they all seem very unlikely.

Code completion depends on analysis, and analysis depends on having a valid and up-to-date `package_config.json` file. If something caused the file to be corrupted that could explain the problem.

Analysis also depends on being able to read the source for those packages from the `.pub-cache` directory. If one of the packages in that directory was corrupted that could potentially cause problems.

Or if some of the files needed for analysis are on a file system that became inaccessible, that could potentially cause problems.

> ... I ran android emulator inside tool window ( I ran it as a separate window before.)

@stevemessick Is there something different about how the emulator runs when running in a tool window that would explain this?

@BojanDolic If you manually type `BoxDecoration` does the analyzer find the class (not produce a diagnostic saying that it's undefined)? Can you navigate to it? If the answer to these is "yes" than analysis appears to be working correctly, so my previous guesses are disproven.

**Comment by BojanDolic** [comment]
@bwilkerson Yes, I can even click to open its source code. Here is the video, hope it will help:

https://user-images.githubusercontent.com/63774884/155766366-a1ae5f19-c4ff-4ebc-91cb-9c81b92f01d0.mp4

**Comment by stevemessick** [comment]
@BojanDolic When using the project you re-created and running the emulator in the embedded window, does completion work as expected?

**Comment by BojanDolic** [comment]
> @BojanDolic When using the project you re-created and running the emulator in the embedded window, does completion work as expected?

I'm running the emulator as a separate window, I have a fear to run it inside the tool window again because the project grew and I don't want to lose it. Completion works fine in the new project so far.

**I tried to recreate the problem in the new project but I couldn't.** I tried running the emulator inside the tool window, getting dependencies, but none of that worked. 

This is not the first time this happened to me, I was watching a tutorial a couple of weeks ago and I cloned the GitHub repo for the tutorial and after a day or two same thing happened but I was still learning so I thought I messed up something and the project was pretty old in terms of flutter version so I finished that project without completion because it was not big.

**Comment by stevemessick** [comment]
I'm sorry @BojanDolic but I didn't quite follow that. If you are concerned about possibly corrupting your project then I'd suggest making a zip file of it, then trying the embedded emulator. If completion gets broken you can delete the project and unzip the zip file to restore.

I can't think of any way using the embedded emulator would affect completion. That's why I'm asking if you can confirm that completion does work with the embedded emulator.

Also, what's the tutorial you were using when you had this problem before?

**Comment by BojanDolic** [comment]
@stevemessick Okay I ran the emulator inside the tool window and it seems that everything works fine, so I'm really confused about what could stop auto-completion to work.


> Also, what's the tutorial you were using when you had this problem before?

I was following a course by Angela Yu about flutter, here is the link to the course:
[Course](https://www.udemy.com/course/flutter-bootcamp-with-dart/)

**Comment by stevemessick** [comment]
Thanks for the link. Apparently, something can get mis-configured in the project that causes completion to stop working correctly. Completion is complex; it is actually implemented in the Dart plugin, which uses the Dart analysis server to do the heavy lifting of computing suggestions, but also provides a hook that the Flutter plugin uses to support its outline. So, in theory, something could go wrong in any of the three components (Dart plugin, Flutter plugin, Dart analyzer).

It sounds like you are in good shape now. If this happens again, please let us know. For now, I'm going to keep it open and hope I get some time to investigate the tutorial you referenced.

One more thing. I mentioned `flutter pub upgrade` earlier, and I see @bwilkerson mentioned a couple symptoms related to that. My hope was that `flutter pub upgrade` would rebuild the pub cache, but you may need to do `flutter clean` first. If you still have the broken project, could you try doing `Flutter Clean` then `Flutter Pub Upgrade` -- they are both on the `Tools > Flutter` nested menu. Then let us know if anything changed. I appreciate your help tracking this down.

**Comment by BojanDolic** [comment]
@stevemessick Thanks for suggestion, I tried `Flutter Clean` and `Flutter Pub Upgrade` and nothing changed, completion still does not work. 

If I recall correctly when I opened the emulator in a tool window I also added the image as an asset so I had to change `pubspec.yaml` and run `pub get`. Maybe that could also point to something. 
I imported assets before that but I didn't run the emulator inside the tool window so I thought the emulator caused this issue but apparently it didn't. Hope this helps.

**Comment by stevemessick** [comment]
Thanks for the reply. I can't help thinking something is wrong in your pub cache. But aside from completely deleting it I don't know how to determine if that is true. I don't recommend deleting it because it is shared among all Flutter and Dart projects. I'm out of ideas, I'm afraid. At least we got you back to a workable state.

**Comment by rorystephenson** [comment]
I think I have the same problem and I've put up with it for quite a while now (between half a year and a year I guess) but it would be great to find a solution.

I blew away the pub cache, ./dart_tool and the flutter/dart plugins in Android Studio with no long. All plugins and Android Studio are up to date.

I'd be up for helping out debugging if there are some log files I can look in to or something else.

---

To be clear: if I type BoxDecoration there is no completion suggested and I can't auto-import. I can navigate to it though.

**Comment by rorystephenson** [comment]
As a last resort I tried using `git bisect` to determine if it was a non-config related change and it was. For some reason quite a while ago I removed flutter as an explicit dependency in my pubspec.yaml 🤦 . Surprisingly everything continued to work except for auto-completions and running flutter tests via right-click.

Adding the following back has made everything work again, hopefully this solves it for someone else too:

```yaml
dependencies:
  flutter:
    sdk: flutter
```

**Comment by BojanDolic** [comment]
> As a last resort I tried using `git bisect` to determine if it was a non-config related change and it was. For some reason quite a while ago I removed flutter as an explicit dependency in my pubspec.yaml 🤦 . Surprisingly everything continued to work except for auto-completions and running flutter tests via right-click.
> 
> Adding the following back has made everything work again, hopefully this solves it for someone else too:
> 
> ```yaml
> dependencies:
>   flutter:
>     sdk: flutter
> ```

Thanks for participating and presenting your solution.

Unfortunately, my `pubspec.yaml` file has those lines and autocomplete doesn't work. Probably something different happened with my project :(

**Comment by stevemessick** [comment]
@BojanDolic Re-reading this just now, I thought of one more thing to try, and another question.

First, the question. You mentioned doing `pub get` at one point. Did you do `flutter pub get` (which the IDE automatically does if you use the menu)? II don't recall why `flutter pub get` is different from `pub get` (which is now `dart pub get`), but apparently it is.

And then you could try deleting the `.dart_tool` directory and doing `flutter pub get` to see if that helps.

**Comment by BojanDolic** [comment]
Sorry for the late response.

I was using `pub get` from the `pubspec.yaml` file when I imported library.

I deleted `.dart_tool` directory and ran `flutter pub get` and unfortunately, completion still doesn't work.

**Comment by stevemessick** [comment]
Ah, that would have done `flutter pub get` so it was doing the right thing. Thanks for the update. I'm out of ideas so I'll leave this open for now.

**Comment by drowtram** [comment]
我也遇到了这样的问题，最近在学习Flutter开发，在创建工程时，Model Name命名了一个大写开头的英文单词，然后自动补全在我学习的这几天就一直是这样，无法自动补全，今天无意间把Model Name文件名改成了小写开头的英文单词，然后在再Android Studio里面打开工程，自动补全就好使了。

**Comment by stevemessick** [comment]
According to Google translate, the previous comment is:
> I also encountered such a problem. I was learning Flutter development recently. When creating a project, the Model Name named a capitalized English word, and then auto-completion has been like this for the past few days that I have been studying, and it cannot be auto-completed. , today I accidentally changed the Model Name file name to English words starting with lowercase, and then opened the project in Android Studio, and auto-completion was fine.

That's interesting. I'm not sure what `Model Name` refers to, but based on the capitalization comment I tried changing the name of `main.dart` to `Main.dart` and changed the project name in `pubspec.yaml` to use camel-case. Neither had an effect on completion, though.

@BojanDolic Could your project be affected by capitalization?

**Comment by BojanDolic** [comment]
@stevemessick I opened the project to test capitalization. Before I did anything I wanted to test if autocompletion works by some miracle so I went to my screen file and to my surprise, **autocompletion now WORKS**. I didn't change anything nor did I open the project from the last time.

Before I opened this old project I worked on my current project and updated Kotlin and Dart plugins. I restarted AS and then opened my old project. Is it possible that this somehow fixed it ?

**Comment by ercantomac** [comment]
It recently started to happen to me as well, and I think it was after updating to Flutter 3.0. I hope they fix it with a minor update soon.

**Comment by officialFlutterDeveloper** [comment]
> It recently started to happen to me as well, and I think it was after updating to Flutter 3.0. I hope they fix it with a minor update soon.

same issue

**Comment by thassio-vinicius** [comment]
This just started happening to me after Flutter 3.0, suggestions like setState and overrides such as initState and didChangeDependecies aren't working anymore.

**Comment by officialFlutterDeveloper** [comment]
I second @thassio-vinicius this is a small issue but it lags the productivity of developer and it needs to be fixed

**Comment by cocoa-rum** [comment]
> This just started happening to me after Flutter 3.0, suggestions like setState and overrides such as initState and didChangeDependecies aren't working anymore. May we reopen this please?

Same here. Tried to reset IDE, did not help. 

> Android Studio Chipmunk | 2021.2.1
> Build #AI-212.5712.43.2112.8512546, built on April 28, 2022
> Non-Bundled Plugins: Dart (212.5744), org.jetbrains.kotlin (212-1.6.21-release-334-AS5457.46), io.flutter (68.1.2)

_For ex VS Code works as before_

**Comment by Fltme** [comment]
> This just started happening to me after Flutter 3.0, suggestions like setState and overrides such as initState and didChangeDependecies aren't working anymore. May we reopen this please?

Omg, finally I find out someone who has the same issue with a setState. I'm new to programming (and learning Flutter) and can't realize the reason for such thing. So, this is just a glitch of 3.0 and will be fixed? ;)

**Comment by igoriuz** [comment]
> > This just started happening to me after Flutter 3.0, suggestions like setState and overrides such as initState and didChangeDependecies aren't working anymore. May we reopen this please?
> 
> Omg, finally I find out someone who has the same issue with a setState. I'm new to programming (and learning Flutter) and can't realize the reason for such thing. So, this is just a glitch of 3.0 and will be fixed? ;)

Should be fixed according to [3.0.2](https://github.com/flutter/flutter/wiki/Hotfixes-to-the-Stable-Channel#302-june-10-2022)

**Comment by thassio-vinicius** [comment]
> > > This just started happening to me after Flutter 3.0, suggestions like setState and overrides such as initState and didChangeDependecies aren't working anymore. May we reopen this please?
> > 
> > 
> > Omg, finally I find out someone who has the same issue with a setState. I'm new to programming (and learning Flutter) and can't realize the reason for such thing. So, this is just a glitch of 3.0 and will be fixed? ;)
> 
> Should be fixed according to [3.0.2](https://github.com/flutter/flutter/wiki/Hotfixes-to-the-Stable-Channel#302-june-10-2022)
> 
> EDIT: Apparently setState still not working [dart-lang/sdk#49027 (comment)](https://github.com/dart-lang/sdk/issues/49027#issuecomment-1152743664)

Yep, setState still not working, can confirm it.

**Comment by ranierjardim** [comment]
https://user-images.githubusercontent.com/12734534/174441369-fc4b4bee-eb9e-472d-9572-fc76500707a1.mov

I'm having problems related to imports too, the auto-complete only works when i force using CTRL + SPACEBAR two times

**Comment by ahmadbintariq4u** [comment]
> > As a last resort I tried using `git bisect` to determine if it was a non-config related change and it was. For some reason quite a while ago I removed flutter as an explicit dependency in my pubspec.yaml 🤦 . Surprisingly everything continued to work except for auto-completions and running flutter tests via right-click.
> > Adding the following back has made everything work again, hopefully this solves it for someone else too:
> > ```yaml
> > dependencies:
> >   flutter:
> >     sdk: flutter
> > ```
> 
> @rorystephenson Thanks for participating and presenting your solution.
> 
> Unfortunately, my `pubspec.yaml` file has those lines and autocomplete doesn't work. Probably something different happened with my project :(

Oh Thanku Man

**Comment by AbelTarazonaApprecio** [comment]
Is there a solution? Same problem. Import, setState, etc

**Comment by AbelTarazonaApprecio** [comment]
Flutter 3.0.3 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 676cefaaff (9 days ago) • 2022-06-22 11:34:49 -0700
Engine • revision ffe7b86a1e
Tools • Dart 2.17.5 • DevTools 2.12.2

**Comment by omidshojaee** [comment]
I have this issue with everything not just initState or setState - with or without running emulator in tools window. This has forced me to use VS Code for now.

```
Android Studio Chipmunk | 2021.2.1 Patch 1
Build #AI-212.5712.43.2112.8609683, built on May 18, 2022
Runtime version: 11.0.12+7-b1504.28-7817840 amd64
VM: OpenJDK 64-Bit Server VM by Oracle Corporation
Windows 10 10.0
GC: G1 Young Generation, G1 Old Generation
Memory: 1280M
Cores: 4
Registry: external.system.auto.import.disabled=true
Non-Bundled Plugins: Dart (212.5744), Docker (212.5712.51), org.jetbrains.kotlin (212-1.7.10-release-333-AS5457.46), io.flutter (69.0.2)
```
```
Flutter plugin 69.0.2
Dart plugin 212.5744
```
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.0.5, on Microsoft Windows [Version 10.0.22000.795], locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✗] Chrome - develop for the web (Cannot find Chrome executable at .\Google\Chrome\Application\chrome.exe)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
[✗] Visual Studio - develop for Windows
    ✗ Visual Studio not installed; this is necessary for Windows development.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components
[✓] Android Studio (version 2021.2)
[✓] VS Code (version 1.70.0-insider)
[✓] Connected device (3 available)
[✓] HTTP Host Availability

! Doctor found issues in 2 categories.
```

**Comment by AlexeyRottCr** [comment]
I often encounter this problem...

**Comment by callmejm** [comment]
I facing this issue after I upgrade my vscode

**Comment by AbelTarazona** [comment]
Same problem... Android Studio Electric Eel | 2022.1.1 and Flutter 3.7.0

Suggestions not working but autocomplete works using CTRL + SPACEBAR two times

**Comment by geronimol** [comment]
Same problem:
Android Studio Hedgehog | 2023.1.1 Patch 
Flutter 3.19
